### ISSUE TITLE
Review the fixes for #144 now that a Roslyn bug has been fixed

### ISSUE THREAD

**Issue by metoule** [dynamicexpresso/DynamicExpresso#184]
There was a bug in the Roslyn compiler (see [here](https://github.com/dotnet/roslyn/issues/53478) and [here](https://github.com/dotnet/roslyn/issues/51518)) that prevented the default value of a parameter from being populated when getting the `ParameterInfo` of local functions / delegates:

```csharp
static bool GetGFunction2(string arg = null)
{
	return arg != null;
}

GFunction gFunc2 = GetGFunction2;
Assert.False(gFunc2.Method.GetParameters()[0].HasDefaultValue); // should be true!
```

That bug [has been fixed in C# 10](https://github.com/dotnet/roslyn/pull/53402). 

This has a few impacts:

- the tests I added for #144  are failing on VS2022 when compiling for .NET5
- we might be able to simplify the way the `SetFunction` method works

**Comment by davideicardi** [comment]
@metoule Thank you for the investigation.

To be compatible with multiple .NET versions we should use some conditional compilation? `#if DOTNET5` ?

**Comment by metoule** [comment]
@davideicardi the issue only occurs for local functions, which was introduced in C# 8 (so netstandard 2.1 and .NET Core 3.x). There's already a conditional compilation around the unit tests `#if NETCOREAPP2_1_OR_GREATER`, which I now realize should be `#if NETCOREAPP3_0_OR_GREATER` :D

I don't think there will be a need for conditional compilation, because the fix is not linked to the runtime, but rather the compiler itself. Which makes it tricky: if I fix the test for the new SDK (VS2022), it will fail with the old SDK (VS2019)!

It also means that I might not be able to change the way `ParseMethodGroupInvocation` works, because the issue might still occur for users that use DynamicExpresso with VS2019..

I'm not really sure what to do for now.

**Comment by davideicardi** [comment]
Maybe we can start by creating a branch for `VS2022` and then we can eval what are exactly the differences and understand what is the best strategy?

**Comment by metoule** [comment]
Good idea! Though for now the only impact I see is those unit tests :)

**Comment by metoule** [comment]
I'm closing this issue, because it's no longer relevant: the test cases were based on SetFunction allowing the registration of multiple delegates with the same signature, but that will no longer be possible after #231
